core/sys/stack-checker.c

```diff
@@ -0,0 +1,104 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the Institute nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ */
+
+/**
+ * \addtogroup stack
+ * @{
+ */
+
+/**
+ * \file
+ * Implementation of the stack checker library.
+ * \author
+ * Atis Elsts <atis.elsts@bristol.ac.uk>
+ */
+
+#if CONTIKI_TARGET_Z1 || CONTIKI_TARGET_SKY | CONTIKI_TARGET_WISMOTE || CONTIKI_TARGET_EXP5438 || CONTIKI_TARGET_SRF06_CC26XX
+
+#include "sys/cc.h"
+#include "sys/stack-checker.h"
+#include "dev/watchdog.h"
+
+/*---------------------------------------------------------------------------*/
+/* linker will provide symbol for end-of-memory */
+extern uint8_t _end;
+/*---------------------------------------------------------------------------*/
+/* The symbol with which the stack memory are is initially fiilled */
+#define STACK_FILL 0xcdcd
+/*---------------------------------------------------------------------------*/
+/* Round up a value to the nearest even number */
+#define ROUND_UP(a) (((a) + 1) & 0xfffffffe)
+/*---------------------------------------------------------------------------*/
+void
+stack_checker_init(void)
+{
+  uint16_t local;
+  uint16_t *p = &local;
+
+  /* printf("eoh=%p eos=%p stack=%p\n", &_end, STACK_ORIGIN, p); */
+
+  while(p > (uint16_t *)&_end) {
+    *p-- = STACK_FILL;
+    /* Make sure WD is not trigerred */
+    watchdog_periodic();
+  }
+}
+/*---------------------------------------------------------------------------*/
+uint16_t
+stack_checker_get_usage(void)
+{
+  uint16_t *p = (uint16_t *)ROUND_UP((unsigned int)&_end);
+  /* skip the bytes used after heap; it's 1 byte by default for _end_of_heap,
+   * more than that means dynamic memory allocation is used somewhere.
+   */
+  while(*p != STACK_FILL) {
+    p++;
+    /* Make sure WD is not trigerred */
+    watchdog_periodic();
+  }
+
+  while(*p == STACK_FILL) {
+    p++;
+    /* Make sure WD is not trigerred */
+    watchdog_periodic();
```

Comment:
>It would be faster to "pet" the watchdog _before_ the first loop, and _after_ the second loop.

Hmpf!!  Maybe not -- some watchdogs are set for only one second -- code must placate them very often!